modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/batch/engine/ObjectFieldColumnDescriptorsImpl.java

```diff
@@ -0,0 +1,278 @@
+				objectFieldName, objectField.getListTypeDefinitionId(), index,
+				objectField.getBusinessType(), propertiesObjectValuePair);
+		}
+
+		if (Objects.equals(
+				objectField.getBusinessType(),
+				ObjectFieldConstants.BUSINESS_TYPE_PICKLIST)) {
+
+			return _getPickListColumnDescriptors(
+				objectFieldName, index, objectField.getBusinessType(),
+				propertiesObjectValuePair);
+		}
+
+		return new ColumnDescriptor[] {
+			ColumnDescriptor.from(
+				objectFieldName, index,
+				_getObjectEntryCustomFieldUnsafeFunction(
+					objectField.getBusinessType(), propertiesObjectValuePair,
+					objectFieldName))
+		};
+	}
+
+	private String _getListEntryValue(Object object, String fieldName) {
+		ListEntry listEntry = (ListEntry)object;
+
+		if (Objects.equals(fieldName, "key")) {
+			return listEntry.getKey();
+		}
+
+		return listEntry.getName();
+	}
+
+	private String _getMultiselectListEntryValue(
+		List<ListEntry> listEntries, String fieldName) {
+
+		String[] parts = fieldName.split(StringPool.UNDERLINE);
+
+		int columnIndex = GetterUtil.getInteger(parts[1]);
+
+		if (listEntries.size() < columnIndex) {
+			return StringPool.BLANK;
+		}
+
+		ListEntry listEntry = listEntries.get(columnIndex - 1);
+
+		if (Objects.equals(parts[0], "key")) {
+			return listEntry.getKey();
+		}
+
+		return listEntry.getName();
+	}
+
+	private ColumnDescriptor[] _getMultiselectPickListColumnDescriptors(
+		String fieldName, long listTypeDefinitionId, int index,
+		String objectFieldBusinessType,
+		ObjectValuePair<Field, Method> propertiesObjectValuePair) {
+
+		int listTypeEntriesCount =
+			_listTypeEntryLocalService.getListTypeEntriesCount(
+				listTypeDefinitionId);
+
+		ColumnDescriptor[] multiselectPickListColumnDescriptors =
+			new ColumnDescriptor[listTypeEntriesCount * 2];
+
+		int listTypeEntriesHeaderIndex = 1;
+
+		for (int i = 0; i < multiselectPickListColumnDescriptors.length;
+			 i = i + 2) {
+
+			multiselectPickListColumnDescriptors[i] = ColumnDescriptor.from(
+				StringBundler.concat(
+					fieldName, ".key_", listTypeEntriesHeaderIndex),
+				index++,
+				_getObjectEntryCustomFieldUnsafeFunction(
+					objectFieldBusinessType, propertiesObjectValuePair,
+					fieldName, "key_" + listTypeEntriesHeaderIndex));
+
+			multiselectPickListColumnDescriptors[i + 1] = ColumnDescriptor.from(
+				StringBundler.concat(
+					fieldName, ".name_", listTypeEntriesHeaderIndex),
+				index++,
+				_getObjectEntryCustomFieldUnsafeFunction(
+					objectFieldBusinessType, propertiesObjectValuePair,
+					fieldName, "name_" + listTypeEntriesHeaderIndex));
+
+			listTypeEntriesHeaderIndex++;
+		}
+
+		return multiselectPickListColumnDescriptors;
+	}
+
+	private UnsafeFunction<Object, Object, ReflectiveOperationException>
+		_getObjectEntryCustomFieldUnsafeFunction(
+			String objectFieldBusinessType,
+			ObjectValuePair<Field, Method> propertiesObjectValuePair,
+			String... fieldNames) {
+
+		if (Objects.equals(
```

Comment:
Maybe we can simplify this like:

```java
return new UnsafeFunction
	<Object, Object, ReflectiveOperationException>() {

	@Override
	public Object apply(Object object)
		throws ReflectiveOperationException {

		Map<?, ?> map = (Map<?, ?>)_getValue(
			object, propertiesObjectValuePair);

		Object value = map.get(fieldNames[0]);

		if (value == null) {
			return StringPool.BLANK;
		}

		if (Objects.equals(
			objectFieldBusinessType,
			ObjectFieldConstants.BUSINESS_TYPE_MULTISELECT_PICKLIST)) {

			return _getMultiselectListEntryValue(
				(List<ListEntry>)value, fieldNames[1]);

		} else if (Objects.equals(
			objectFieldBusinessType,
			ObjectFieldConstants.BUSINESS_TYPE_PICKLIST)) {

			return _getListEntryValue(value, fieldNames[1]);
		}

		return CSVUtil.encode(value);
	}

};
```

What do you think?

Reply:
I agree this looks tempting to do, but the idea of these UnsafeFunctions is to assign a specific function, that will extract value from the field, to a specific ObjectEntry field. That way we don't have to run through all if conditions on **each** ObjectEntry export, instead we already have value extracting logic ready.